src/main/java/pl/joegreen/sergeants/simulator/GameMap.java

```diff
@@ -0,0 +1,240 @@
+     */
+    private int[] getCitiesDiff() {
+        List<Integer> allCities = Arrays.stream(tiles)
+                .filter(tile -> tile.getClass() == City.class)
+                .map(Tile::getTileIndex)
+                .collect(Collectors.toList());
+
+        allCities.add(0, allCities.size()); //second element = wrong
+        allCities.add(0, 0);//first element = zero correct
+        return allCities.stream().mapToInt(i -> i).toArray();
+    }
+
+    /**
+     * Returns all generals, visible and fogged. This is not a problem since bot will still receive tile as fogged (-3).
+     *
+     * @return list of tileIndex where general is, array index is player index
+     */
+    private int[] getGenerals() {
+        return Arrays.stream(tiles)
+                .filter(tile1 -> tile1.getClass().equals(General.class))
+                .sorted((t1, t2) -> t1.getPlayerIndex() - t2.getPlayerIndex())
+                .mapToInt(Tile::getTileIndex)
+                .toArray();
+    }
+
+    private boolean isVisible(int playerIndex, Tile tile) {
+        boolean selfTile = tile.getPlayerIndex() == playerIndex;
+        return selfTile || getSurroundingTiles(tile).stream().anyMatch(t -> t.getPlayerIndex() == playerIndex);
+    }
+
+    private List<Tile> getSurroundingTiles(Tile tile) {
+        //this could be prettier
+        List<Tile> ret = new ArrayList<>();
+        getTileLeftOf(tile).ifPresent(leftTile -> {
+            ret.add(leftTile);
+            getTileAbove(leftTile).ifPresent(ret::add);
+            getTileBelow(leftTile).ifPresent(ret::add);
+        });
+        getTileRightOf(tile).ifPresent(t -> {
+            ret.add(t);
+            getTileAbove(t).ifPresent(ret::add);
+            getTileBelow(t).ifPresent(ret::add);
+        });
+        getTileAbove(tile).ifPresent(ret::add);
+        getTileBelow(tile).ifPresent(ret::add);
+        return ret;
+    }
+
+    private Optional<Tile> getTileAbove(Tile tile) {
+        int tileIndex = tile.getTileIndex() - width;
+        return tileIndex >= 0 ? Optional.of(tiles[tileIndex]) : Optional.empty();
+    }
+
+    private Optional<Tile> getTileBelow(Tile tile) {
+        int tileIndex = tile.getTileIndex() + width;
+        return tileIndex < tiles.length ? Optional.of(tiles[tileIndex]) : Optional.empty();
+    }
+
+    private Optional<Tile> getTileRightOf(Tile tile) {
+        boolean isMostRight = tile.getTileIndex() % width == width - 1;
+        return isMostRight ? Optional.empty() : Optional.of(tiles[tile.getTileIndex() + 1]);
+    }
+
+    private Optional<Tile> getTileLeftOf(Tile tile) {
+        int tileIndex = tile.getTileIndex();
+        boolean isMostLeft = tileIndex == 0 || tileIndex % width == 0;
+        return isMostLeft ? Optional.empty() : Optional.of(tiles[tileIndex - 1]);
+    }
+
+    public int getHalfTurnCounter() {
+        return halfTurnCounter;
+    }
+
+    public void saveGameAsJson() {
+        //pls don't look at this code
```

Comment:
To pass a filehandler sounds like a good idea. So we dont merge but I will still keep it for debugging until we get a proper solution in place. I think a good way is to keep this in another branch or a separate class.